syntax = "proto3";
package dcq.v1;

option go_package = "dcq.dev/api/dcqv1";
option java_multiple_files = true;

message Empty {}
message Ack { bool ok = 1; string msg = 2; }

enum Domain { DOMAIN_UNKNOWN = 0; FSO = 1; MMWAVE = 2; LEO = 3; WIFI7 = 4; FR3_6G = 5; }
enum SloClass { SLO_UNKNOWN = 0; URLLC = 1; EMBB = 2; BESTEFFORT = 3; }

message HelloReq { string plugin_name = 1; string version = 2; string git_sha = 3; }
message HelloResp { string bridge_version = 1; string qcs_firmware = 2; repeated string features = 3; }

message Capabilities {
  bool can_plan_tx_schedule = 1;
  bool can_phase_dither = 2;
  bool can_clock_align = 3;
  bool can_domain_policy = 4;
  bool requires_raw_counts = 5;
}

message ClockModel {
  double coarse_ppb = 1;
  double fine_hz = 2;
  double tdc_bin_ps = 3;
  double gate_ns = 4;
}

message Telemetry {
  int64 t_unix_ms = 1;
  double qber_pct = 2;
  double sifted_rate_cps = 3;
  double secure_rate_bps = 4;
  double jitter_ps = 5;
  double atm_loss_db_per_km = 6;
  double dark_cps = 7;
  double det_eff = 8;
  double temperature_c = 9;
  string site = 10;
  Domain active_domain = 11;
  double scintillation_idx = 12;
}

message Constraints {
  double mu_min = 1;
  double mu_max = 2;
  double rep_rate_min_hz = 3;
  double rep_rate_max_hz = 4;
  double qber_hard_ceiling_pct = 5;
}

message Slo {
  SloClass cls = 1;
  double jitter_ps_target = 2;
  double key_rate_min_bps = 3;
}

message PlanRequest {
  ClockModel clock = 1;
  Telemetry tel = 2;
  Constraints limits = 3;
  Slo slo = 4;
}

message DecoyProfile {
  double mu_signal = 1;
  double mu_decoy = 2;
  double vac_prob = 3;
  double sig_prob = 4;
  double decoy_prob = 5;
}

message TxOverrides {
  double rep_rate_hz = 1;
  double pulse_width_ps = 2;
  DecoyProfile decoys = 3;
  double gate_shift_ps = 4;
}

message PhaseOverrides {
  double amzi_phase_deg = 1;
  double eom_bias_v_delta = 2;
}

message DomainPolicy {
  Domain preferred = 1;
  string srv6_bsid = 2;
  int32 dscp = 3;
  bool mlo_prefer_6ghz = 4;
}

message PlanResponse {
  TxOverrides tx = 1;
  PhaseOverrides phase = 2;
  DomainPolicy domain = 3;
  uint32 next_cycle_ms = 4;
  string rationale = 5;
}

service DualClockPlugin {
  rpc Hello(HelloReq) returns (HelloResp);
  rpc Describe(Empty) returns (Capabilities);
  rpc SetClockModel(ClockModel) returns (Ack);
  rpc PlanCycle(PlanRequest) returns (PlanResponse);
  rpc Events(stream Telemetry) returns (Ack);
}
